### ISSUE TITLE
Improved log formatting (part 2)

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal#442]
Following up on this issue: https://github.com/fractal-analytics-platform/fractal-server/issues/501

The log formatting still is suboptimal, at least in some cases. The output of `fractal task check-collection 1` when we have a collection failure looks like this:
<img width="1340" alt="Screenshot 2023-02-24 at 10 38 09" src="https://user-images.githubusercontent.com/18033446/221144777-bb236e6f-8552-4b00-8d1f-554d4551b73d.png">

Wasn't fractal-analytics-platform/fractal-server#501 supposed to handle the `\n` formatting?

**Comment by tcompa** [comment]
As of #443, the check-collection command should return an output like
```
{
  "id": 1,
  "data": {
    "status": "OK",
    "package": "fractal_tasks_dummy",
    "venv_path": ".fractal/fractal_tasks_dummy0.1.0",
    "task_list": [
      {
        "name": "dummy",
        "source": "pip-local:fractal_tasks_dummy-0.1.0-py3-none-any.whl",
        "id": 1,
        "command": "/tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/python /tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/lib/python3.10/site-packages/fractal_tasks_dummy/dummy.py",
        "input_type": "Any",
        "output_type": "Any",
        "default_args": {
          "message": "default message",
          "index": 0,
          "raise_error": false
        },
        "meta": {}
      },
      {
        "name": "dummy parallel",
        "source": "pip-local:fractal_tasks_dummy-0.1.0-py3-none-any.whl",
        "id": 2,
        "command": "/tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/python /tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/lib/python3.10/site-packages/fractal_tasks_dummy/dummy_parallel.py",
        "input_type": "Any",
        "output_type": "Any",
        "default_args": {
          "message": "default message",
          "raise_error": false
        },
        "meta": {
          "parallelization_level": "index"
        }
      }
    ],
    "info": null
  },
  "timestamp": "2023-02-24 10:24:57.622550"
}

This is the task-collection log:

Creating venv and installing package
Subprocess call to: /tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/pip install --upgrade pip
Requirement already satisfied: pip in ./venv/lib/python3.10/site-packages (22.0.2)
Collecting pip
  Using cached pip-23.0.1-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 22.0.2
    Uninstalling pip-22.0.2:
      Successfully uninstalled pip-22.0.2
Successfully installed pip-23.0.1


Subprocess call to: /tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/pip install /home/tommaso/Fractal/fractal/tests/data/fractal_tasks_dummy-0.1.0-py3-none-any.whl
Processing /home/tommaso/Fractal/fractal/tests/data/fractal_tasks_dummy-0.1.0-py3-none-any.whl
Collecting pydantic<2.0.0,>=1.10.1
  Using cached pydantic-1.10.5-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)
Collecting typing-extensions>=4.2.0
  Using cached typing_extensions-4.5.0-py3-none-any.whl (27 kB)
Installing collected packages: typing-extensions, pydantic, fractal-tasks-dummy
Successfully installed fractal-tasks-dummy-0.1.0 pydantic-1.10.5 typing-extensions-4.5.0


Subprocess call to: /tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/pip show fractal_tasks_dummy
Name: fractal-tasks-dummy
Version: 0.1.0
Summary: 
Home-page: 
Author: Jacopo Nespolo
Author-email: jacopo.nespolo@exact-lab.it
License: 
Location: /tmp/pytest-of-tommaso/pytest-144/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/lib/python3.10/site-packages
Requires: pydantic
Required-by: 


loading manifest
manifest loaded
```

This is for a successful collection, but the same formatting should apply to all cases - including failures

**Comment by tcompa** [comment]
> Can "pretty log formatting" be added to all fractal client commands that can return error messages?

I  think we only have these two (`task check-collection` and `job show`), but let's open new issues if needed.

**Comment by jluethi** [comment]
Do we want to always show this extensive log, even in the success case? There, the status and task list would be sufficient in normal cases (with an option to show log if so desired)

**Comment by tcompa** [comment]
> Do we want to always show this extensive log, even in the success case? There, the status and task list would be sufficient in normal cases (with an option to show log if so desired)

This was the log with `--include-logs`. Without any flag, it looks like
```
{
  "id": 1,
  "data": {
    "status": "OK",
    "package": "fractal_tasks_dummy",
    "venv_path": ".fractal/fractal_tasks_dummy0.1.0",
    "task_list": [
      {
        "name": "dummy",
        "meta": {},
        "source": "pip-local:fractal_tasks_dummy-0.1.0-py3-none-any.whl",
        "command": "/tmp/pytest-of-tommaso/pytest-145/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/python /tmp/pytest-of-tommaso/pytest-145/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/lib/python3.10/site-packages/fractal_tasks_dummy/dummy.py",
        "output_type": "Any",
        "default_args": {
          "message": "default message",
          "index": 0,
          "raise_error": false
        },
        "id": 1,
        "input_type": "Any"
      },
      {
        "name": "dummy parallel",
        "meta": {
          "parallelization_level": "index"
        },
        "source": "pip-local:fractal_tasks_dummy-0.1.0-py3-none-any.whl",
        "command": "/tmp/pytest-of-tommaso/pytest-145/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/python /tmp/pytest-of-tommaso/pytest-145/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/lib/python3.10/site-packages/fractal_tasks_dummy/dummy_parallel.py",
        "output_type": "Any",
        "default_args": {
          "message": "default message",
          "raise_error": false
        },
        "id": 2,
        "input_type": "Any"
      }
    ],
    "log": null,
    "info": null
  },
  "timestamp": "2023-02-24 10:30:32.224432"
}
```

**Comment by jluethi** [comment]
That's great!

Why does it say `"log": null,` in the output in the case without the `--include-logs` flag?

And do we automatically get the log in the failure case?

**Comment by tcompa** [comment]
> Why does it say "log": null, in the output in the case without the --include-logs flag?

This is the response of the server. If you don't ask for logs when calling the server endpoint, you get a `None`. We can filter the `null` values, if that looks more clear.

> And do we automatically get the log in the failure case?

At the moment, we don't.
You only get logs if you ask the server for logs.
Would you rather have an if/else that checks the status and - if needed - issues a second call to the server?

**Comment by jluethi** [comment]
> This is the response of the server. If you don't ask for logs when calling the server endpoint, you get a None. We can filter the null values, if that looks more clear.

Nice idea. Avoids people thinking no log exists.

> You only get logs if you ask the server for logs.

Makes sense

> Would you rather have an if/else that checks the status and - if needed - issues a second call to the server?

Hmm, good question. Not fully sure. Let's leave it as is for the moment and point people to the `--include-logs` flags were necessary. The log can get quite overwhelming for someone who isn't used to that, and someone who can handle the logs can probably also handle setting the flag :)
Especially if we remove the logs: None, we avoid confusion about "Oh, it failed but there were no logs".

**Comment by tcompa** [comment]
Now the check after a successfull install (and with no flags) produces the following output (without `null` fields):
```
{
  "id": 1,
  "data": {
    "status": "OK",
    "package": "fractal_tasks_dummy",
    "venv_path": ".fractal/fractal_tasks_dummy0.1.0",
    "task_list": [
      {
        "default_args": {
          "message": "default message",
          "index": 0,
          "raise_error": false
        },
        "name": "dummy",
        "id": 1,
        "input_type": "Any",
        "source": "pip-local:fractal_tasks_dummy-0.1.0-py3-none-any.whl",
        "meta": {},
        "command": "/tmp/pytest-of-tommaso/pytest-154/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/python /tmp/pytest-of-tommaso/pytest-154/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/lib/python3.10/site-packages/fractal_tasks_dummy/dummy.py",
        "output_type": "Any"
      },
      {
        "default_args": {
          "message": "default message",
          "raise_error": false
        },
        "name": "dummy parallel",
        "id": 2,
        "input_type": "Any",
        "source": "pip-local:fractal_tasks_dummy-0.1.0-py3-none-any.whl",
        "meta": {
          "parallelization_level": "index"
        },
        "command": "/tmp/pytest-of-tommaso/pytest-154/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/bin/python /tmp/pytest-of-tommaso/pytest-154/test_task_collection_and_list0/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_dummy0.1.0/venv/lib/python3.10/site-packages/fractal_tasks_dummy/dummy_parallel.py",
        "output_type": "Any"
      }
    ]
  },
  "timestamp": "2023-02-24 10:50:10.825735"
}
```

**Comment by jluethi** [comment]
Awesome! 🎉

**Comment by tcompa** [comment]
And we agreed to postpone the automatic inclusion of logs (for status=fail), for the moment.

**Comment by jluethi** [comment]
> And we agreed to postpone the automatic inclusion of logs (for status=fail), for the moment.

Agreed. Not sure whether we'd actually want that.
